x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/use_prebuilt_rules_upgrade_state.ts

```diff
@@ -0,0 +1,116 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useMemo, useState } from 'react';
+import {
+  ThreeWayDiffConflict,
+  type DiffableAllFields,
+  type DiffableRule,
+  type RuleObjectId,
+  type RuleSignatureId,
+  type RuleUpgradeInfoForReview,
+} from '../../../../../../common/api/detection_engine';
+import { convertRuleToDiffable } from '../../../../../../common/detection_engine/prebuilt_rules/diff/convert_rule_to_diffable';
+
+export interface RuleUpgradeState extends RuleUpgradeInfoForReview {
+  /**
+   * Rule containing desired values users expect to see in the upgraded rule.
+   */
+  finalRule: DiffableRule;
+  /**
+   * Indicates whether there are conflicts blocking rule upgrading.
+   */
+  hasUnresolvedConflicts: boolean;
+}
+export type RulesUpgradeState = Record<RuleSignatureId, RuleUpgradeState>;
+export type SetFieldResolvedValueFn<
+  FieldName extends keyof DiffableAllFields = keyof DiffableAllFields
+> = (params: {
+  ruleId: RuleObjectId;
+  fieldName: FieldName;
+  resolvedValue: DiffableAllFields[FieldName];
+}) => void;
+
+type RuleResolvedConflicts = Partial<DiffableAllFields>;
+type RulesResolvedConflicts = Record<string, RuleResolvedConflicts>;
+
+interface UseRulesUpgradeStateResult {
+  rulesUpgradeState: RulesUpgradeState;
+  setFieldResolvedValue: SetFieldResolvedValueFn;
+}
+
+export function usePrebuiltRulesUpgradeState(
+  ruleUpgradeInfos: RuleUpgradeInfoForReview[]
+): UseRulesUpgradeStateResult {
+  const [rulesResolvedConflicts, setRulesResolvedConflicts] = useState<RulesResolvedConflicts>({});
+  const setFieldResolvedValue = useCallback((...[params]: Parameters<SetFieldResolvedValueFn>) => {
+    setRulesResolvedConflicts((prevRulesResolvedConflicts) => ({
+      ...prevRulesResolvedConflicts,
+      [params.ruleId]: {
+        ...(prevRulesResolvedConflicts[params.ruleId] ?? {}),
+        [params.fieldName]: params.resolvedValue,
+      },
+    }));
+  }, []);
+  const rulesUpgradeState = useMemo(() => {
+    const state: RulesUpgradeState = {};
+
+    for (const ruleUpgradeInfo of ruleUpgradeInfos) {
+      state[ruleUpgradeInfo.rule_id] = {
+        ...ruleUpgradeInfo,
+        finalRule: calcFinalDiffableRule(
+          ruleUpgradeInfo,
+          rulesResolvedConflicts[ruleUpgradeInfo.id] ?? {}
+        ),
+        hasUnresolvedConflicts:
+          calcUnresolvedConflicts(
+            ruleUpgradeInfo,
+            rulesResolvedConflicts[ruleUpgradeInfo.id] ?? {}
+          ) > 0,
+      };
+    }
+
+    return state;
+  }, [ruleUpgradeInfos, rulesResolvedConflicts]);
+
+  return {
+    rulesUpgradeState,
+    setFieldResolvedValue,
+  };
+}
+
+function calcFinalDiffableRule(
+  ruleUpgradeInfo: RuleUpgradeInfoForReview,
+  ruleResolvedConflicts: RuleResolvedConflicts
+): DiffableRule {
+  return {
+    ...convertRuleToDiffable(ruleUpgradeInfo.target_rule),
```

Comment:
I'm not sure if this is WIP, but if we want to get the "final version" of a rule, we should get all diffable fields, and then overwrite all fields that have a `merged_version` within `diff.fields`.

The logic here implemented is correct, but it will only work when the user manually does a change in the edit component.

I think the return value of this method should look something like:
```ts
  return {
    ...convertRuleToDiffable(ruleUpgradeInfo.target_rule),
    ...getFieldsWithMergedVersions(ruleUpgradeInfo.diff), // loops through diff.fields and retrieves fields with a merged_version
    ...ruleResolvedConflicts,
  } as DiffableRule;
```

Reply:
This will make sure to use the `target` version as the "inital" diffable rule, then overwrite with any values that were returned with a `merged_version` from /upgrade/_review, and finally overwrite any values that have been manually edited by the user.

Reply:
Thanks. It makes sense. I made proper adjustments in the code.